### ISSUE TITLE
Testing for the pragmatic

### ISSUE THREAD

**Issue by orta** [orta/life#15]
Key things I have had to learn so far.
- What is Dependency Injection.
- The less you have in view controllers at all the easier things are to test.
- If it's public on a class it should have a test
- Unit Tests are
  - Return checking
  - Internal state checking
  - Did someone else get called checking

They are in this order of simplicity.
Screw trying to be fancy and getting 100% coverage.

Dependency Injection basically means moving things that are hard constants into method parameters of instance variables on a class. I'm not at a position to understand the usage of frameworks for this yet.

**Comment by orta** [comment]
> Struggled to test a UITableViewController subclass for anything other than public actions like "save" / "cancel"
- Switched to moving it's fetch requests to be based off a dep injected MOC 
- Then could test number of cells with mockMOCs set up
- This then became the test of internal state: _does it hide artists with published works?_ OR _do I check that x is hidden_ etc

**Comment by jasperblues** [comment]
Certainly agree that you don't need a framework for DI, however we find its nice to put some "rails" around it. 

Three types of DI:
1. Initializer injection
2. Property injection
3. A custom method interface that sets internal state (not really that common)

The other aspect of DI is this: 
- If you keep doing DI, you end up with a top-level "assembly" class that defines components and how they wire together. This way if you need to change one component for another, you only have to change one place, rather than going through all your code looking for instances of that object.

**Comment by jasperblues** [comment]
Here's another approach, which is sometimes good: 

http://sharpfivesoftware.com/2013/03/20/dependency-injection-is-not-a-virtue-in-objective-c/ 

(in most cases I would use DI though).

**Comment by orta** [comment]
Jon Reid's slides from Mobile Conference Europe: https://speakerdeck.com/jonreid/tdd-for-ios-and-anything-by-controlling-dependencies

**Comment by orta** [comment]
#### XCTest vs Specta/Kiwi etc

XCTest is what they all compile to. Solid Objective-C, can get verbose. Methods must start with test. Xcode can find them without it being ran. Essentially is a flat hierarchy of test methods.

```
ARArtworkDownloaderTests
 - (void)testObjectsForObject
 - (void)testOperation
```

Specta / Kiwi et all provide a collection of c functions to make tests shorter and to make a hierarchy where you can have multiple test contexts.

```
AREditAlbumViewControllerTests
context "on init"
    it "accepts an album"
    it "cannot be initialized without an album"

it "has a cell for each available artist

context "nav buttons"
    it "should pop nav on save"
    it "should pop nav on cancel"

```

**Comment by jasperblues** [comment]
I'm using Cedar on my current project. . this particular framework runs as a separate iOS app (though it can also run on XCTest/OCUnit). . .The main driver for making the switch was I had a strange problem where XCTest simply stopped working from the cmd-line, so made the switch. . . (I think it was related to flushing coverage data, which is nasty on iOS7) 

Cedar's matching library has gone from strength to strength and is really nice and easy to read now.

**Comment by orta** [comment]
Also, on being wary of [going too far.](http://accidentaltechnologist.com/programming/maybe-you-are-doing-too-much-test-driven-development/)

**Comment by rbsgn** [comment]
I'm using Cedar too and I like it concise matcher syntax. Although AppCode doesn't detect Specs target created by Cedar (smooth IDE integration is a must).

Move as much of app logic off view controllers as possible and test it with stub view models.
Prefer screenshot comparison for view related code (view controller is also related to UI).

**Comment by jasperblues** [comment]
> What about forking RuboCop into something like CocoCop?

We were talking about OCLint the other day. . . looks promising. The kinds of checks I’m looking for: 
- Class too long
- Too many method parameters
- Cyclomatic complexity (ie too many dependencies)
- Copy paste detection
- Not documented public method

I tried the various build tools around, but to hit my own personal sweet-spot of convention vs configurable, I created: https://github.com/jasperblues/xcode-gradle  (very early days yet). 

. . planning to an OCLint task in there.
